refactor(client): migrate certification to ts

ShaunSHamilton opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

The reason for the change in `cert-and-project-map` is to provide stronger typing.

ShaunSHamilton committed:
refactor(client): migrate certification to ts

ShaunSHamilton committed:
fix: reduce certMap into projectMap etc.

ShaunSHamilton committed:
fix: rename certMap -> fullCertMap

ShaunSHamilton committed:
fix: cert-and-project-map strict typing

ShaunSHamilton committed:
close but not close enough

ShaunSHamilton committed:
fix: combine project and legacy project maps

ShaunSHamilton committed:
chore: add type assertion safety comment

ShaunSHamilton committed:
fix: add correct types to test mocks

ShaunSHamilton committed:
fix: another test

Sboonny commented on client/src/components/settings/certification.tsx:
```diff
@@ -13,60 +12,41 @@ import ProjectPreviewModal from '../../templates/Challenges/components/project-p
   projectMap,
-  legacyProjectMap
+  legacyProjectMap,
+  fullProjectMap
```
```suggestion
  fullProjectMap,
  ProjectMap,
  LegacyProjectMap,
```

Sboonny commented on client/src/components/settings/certification.tsx:
```diff
@@ -226,9 +204,10 @@ export class CertificationSettings extends Component {
-  renderCertifications = (certName, projectsMap) => {
-    const { t } = this.props;
-    const { certSlug } = first(projectsMap[certName]);
+  type CertName = keyof typeof projectMap | keyof typeof legacyProjectMap;
```
```suggestion
  type CertName = keyof ProjectMap | keyof LegacyProjectMap;
```

Sboonny commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -765,23 +769,48 @@ function getJavaScriptAlgoPath(project: string) {
-    }
-  }
-});
+type ProjectMap = Record<
```
```suggestion
export type ProjectMap = Record<
```

Sboonny commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -765,23 +769,48 @@ function getJavaScriptAlgoPath(project: string) {
+  };
+}, {} as ProjectMap);
+
+type LegacyProjectMap = Record<
```
```suggestion
export type LegacyProjectMap = Record<
```

Sboonny reviewed (commented):
this is confusing 🤣, I have tried to research why, and was confused more 🤣.

Suggestions are for readability, to see where are the types related to, they don't add much feel free to ignore it

Sboonny commented:
oh if PR is in progress, you can move it to draft to avoid confusion.

I think murgush mentioned this on one of the meetings

ShaunSHamilton committed:
remove settings-button.test.js

The test did not handle any changes, and only served as a failing snapshot when projects are added/updated

ShaunSHamilton committed:
refactor: use exported map types

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

Sboonny commented on client/src/components/settings/certification.tsx:
```diff
@@ -13,60 +12,41 @@ import ProjectPreviewModal from '../../templates/Challenges/components/project-p
 
-const certifications = Object.keys(projectMap);
-const legacyCertifications = Object.keys(legacyProjectMap);
+// Safety: projectMap definitely has projectMap keys,
+// and we are only interested in these keys
+const certifications = Object.keys(projectMap) as Array<keyof ProjectMap>;
+// Safety: legacyProjectMap definitely has legacyProjectMap keys,
+// and we are only interested in these keys
+const legacyCertifications = Object.keys(legacyProjectMap) as Array<
+  keyof LegacyProjectMap
+>;
```
What do you think of creating a helper function, that creates the types for the keys?

This way if someone need to type keys of an Object, and don't know the types, they can use the utility function and get the types of the object. I checked the codebase and there are some cases which create `Object.keys() as typeof keys` too, and I am slightly worried that a future dev will struggle with these types 🤔

```suggestion
const objectKeys = <Obj extends Record<string, unknown>, Keys extends (keyof Obj)[]>(obj: Obj): Keys => {
  return Object.keys(obj) as Keys;
}

const certifications = objectKeys(projectMap);
const legacyCertifications = objectKeys(legacyProjectMap);
```

ShaunSHamilton replied:
I would not like a utility function that asserts a generic's type.

Anywhere we have an assertion, we should be justifying it.

Sboonny replied:
Valid point, and I don't think the utility function will stop a bug, all it would do is show keys values when you hover over the objectkeys, which can be done through checking definiton.

![image](https://user-images.githubusercontent.com/88248797/228291281-4df74e06-58dc-458c-9ba2-c2259107f4c8.png)

ShaunSHamilton committed:
merge branch upstream/main into refactor_certification

ShaunSHamilton commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -769,23 +773,52 @@ function getJavaScriptAlgoPath(project: string) {
-const projectMap: Partial<Record<Title, unknown>> = {};
+const certMapWithoutFullStack = Array.of(
+  ...upcomingCertMap,
+  // @ts-expect-error TODO: This did not used to be a type error
```
I cannot explain this one. This error was not here, before I pulled upstream.

So, either TS was going crazy on my machine. Or, we have changed TS config such that this now produces an error.

ShaunSHamilton replied:
Turns out the solution was to just use `[...upcomingCertMap, ...]` instead of `Array.of`

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -40,7 +40,7 @@ const legacyInfosecQaInfosecBase = infoSecBase;
 
```
```suggestion
type CertMap = {
  id: string;
  title: string;
  certSlug: string;
  flag: string;
  projects: Readonly<{
    id: string;
    title: string;
    link: string;
    certSlug: string;
  }[]> | null;
};
```

ShaunSHamilton replied:
The issue here is we lose the strong typing. Considering nothing here is dynamically created, it would be better to find a solution that does not lose us this information.

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -40,7 +40,7 @@ const legacyInfosecQaInfosecBase = infoSecBase;
 // TODO: generate this automatically in a separate file
 // from the md/meta.json files for each cert and projects
-const certMap = [
+const legacyCertMap = [
```
```suggestion
const legacyCertMap: Readonly<CertMap[]> = [
```

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -292,7 +285,18 @@ const certMap = [
-  },
+  }
+] as const;
+const legacyFullStack = {
```
```suggestion
const legacyFullStack: Readonly<CertMap> = {
```

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -292,7 +285,18 @@ const certMap = [
+  // Requirements are other certs and is
+  // handled elsewhere
+} as const;
+const certMap = [
```
```suggestion
const certMap: Readonly<CertMap[]> = [
```

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -754,6 +757,7 @@ const certMap = [
     ]
   }
 ] as const;
+const upcomingCertMap = [] as const;
```
```suggestion
const upcomingCertMap: Readonly<CertMap[]> = [] as const;
```

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -769,23 +773,52 @@ function getJavaScriptAlgoPath(project: string) {
-const projectMap: Partial<Record<Title, unknown>> = {};
+const certMapWithoutFullStack = Array.of(
+  ...upcomingCertMap,
+  // @ts-expect-error TODO: This did not used to be a type error
```
```suggestion
```

naomi-lgbt commented on client/src/resources/cert-and-project-map.ts:
```diff
@@ -769,23 +773,52 @@ function getJavaScriptAlgoPath(project: string) {
-    }
-  }
-});
+// @ts-expect-error TODO: This did not used to be a type error
```
```suggestion
```

naomi-lgbt reviewed (commented):
I leave this up to you, but defining an explicit type resolved the issues for me locally.

ShaunSHamilton committed:
fix: use array instead of Array.of

Sboonny commented:
this PR is using an old version of `react-i18next`, and using the wrong type for `Tfunction` merging it will cause the ci to fail in future PRs

ShaunSHamilton committed:
fix: use i18next TFunction

ShaunSHamilton committed:
merge branch upstream/main into refactor_certification

ojeytonwilliams commented:
Hey, I should be able to give this a proper look next week. Sorry for the delay.
